### PR TITLE
MGMT-8130: Unbind hosts on cluster delete

### PR DESCRIPTION
**Pull request by rollandf**
# Assisted Pull Request

## Description

When a cluster is deleted:
- For each of the hosts:
  - If the Host's InfraEnv is associated to the Cluster (not late binding):
    - Delete the Host
  - If the Host's InfraEnv is not associated to the Cluster (late binding):
    - Unbind the Host

## List all the issues related to this PR

- [x] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @filanov 
/cc @slaviered 
/cc @yevgeny-shnaidman 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by rollandf** [comment]
/hold
Need to add tests

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -919,6 +919,25 @@ func (b *bareMetalInventory) DeregisterClusterInternal(ctx context.Context, para
 		log.Warnf("failed to delete DNS record sets for base domain: %s", cluster.BaseDNSDomain)
 	}
 
+	for _, h := range cluster.Hosts {
+		infraEnv, err2 := common.GetInfraEnvFromDB(b.db, h.InfraEnvID)
```
Comment (filanov) [inline internal/bminventory/inventory.go:923]: Consider adding a helper function for this block.

File: internal/bminventory/inventory.go [hunk 2]
```diff
@@ -919,6 +919,25 @@ func (b *bareMetalInventory) DeregisterClusterInternal(ctx context.Context, para
...
+		if err2 != nil {
+			log.WithError(err2).Errorf("failed to get infra env: %s", h.InfraEnvID)
+			return common.NewApiError(http.StatusInternalServerError, err2)
+		}
+		if infraEnv.ClusterID == *cluster.ID {
+			if err = b.hostApi.UnRegisterHost(ctx, h.ID.String(), h.InfraEnvID.String()); err != nil {
+				log.WithError(err).Errorf("failed to delete host: %s", h.ID.String())
+				return common.NewApiError(http.StatusInternalServerError, err)
```
Comment (filanov) [inline internal/bminventory/inventory.go:931]: Is it always internal error?

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (rollandf) [reply internal/bminventory/inventory.go]: Done

File: internal/bminventory/inventory.go (hunk 2 above)
Comment (rollandf) [reply internal/bminventory/inventory.go:946]: `UnRegisterHost`  returns the DB error.
Should we check for more granularity?

Comment (filanov) [reply internal/bminventory/inventory.go:946]: Maybe let the api  level to handle the error type if it's 500?

Comment (rollandf) [reply internal/bminventory/inventory.go]: I update the code. See if it is what you meant.

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -927,6 +932,31 @@ func (b *bareMetalInventory) DeregisterClusterInternal(ctx context.Context, para
...
+		}
+		if infraEnv.ClusterID == *cluster.ID {
+			if err = b.hostApi.UnRegisterHost(ctx, h.ID.String(), h.InfraEnvID.String()); err != nil {
+				log.WithError(err).Errorf("failed to delete host: %s", h.ID.String())
+				return err
+			}
+			eventgen.SendHostDeregisteredEvent(ctx, b.eventsHandler, cluster.ID, *h.ID, h.InfraEnvID,
+				hostutil.GetHostnameForMsg(h))
+		} else {
+			if err = b.hostApi.UnbindHost(ctx, h, b.db); err != nil {
```
Comment (filanov) [inline internal/bminventory/inventory.go:951]: I think that you could add an even here right?

Comment (rollandf) [reply internal/bminventory/inventory.go:951]: We don't have yet events for bind/unbind.
Can we add it in a follow up PR?

Comment (filanov) [reply internal/bminventory/inventory.go:951]: @nmagnezi didn't you handle this part?

File: internal/cluster/registrar.go [hunk 4]
```diff
@@ -118,7 +118,6 @@ func (r *registrar) DeregisterCluster(ctx context.Context, cluster *common.Clust
 	}
 
 	if txErr = common.DeleteRecordsByClusterID(tx, *cluster.ID, []interface{}{
-		&models.Host{},
```
Comment (slaviered) [inline internal/cluster/registrar.go:121]: The code need to be also merged with the cluster cleaning code permanentClusterDeletion so both will be V2 and Unbound compatible

Comment (rollandf) [reply internal/cluster/registrar.go:121]: Seems that Garbage collector is deleting the host in a different function: https://github.com/openshift/assisted-service/blob/master/internal/garbagecollector/garbageCollector.go#L85

I opened a task to update Garbage collector to delete the infraenv objects also: https://issues.redhat.com/browse/MGMT-8178

**Pull request merged**